step-generation/src/utils/constructInvariantContextFromAnalysis.ts

```diff
@@ -0,0 +1,205 @@
+import {
+  getLabwareDefinitionsByURIForProtocol,
+  getModuleType,
+  getPipetteSpecsV2,
+} from '@opentrons/shared-data'
+
+import { uuid } from '.'
+import { GRIPPER_LOCATION } from '../constants'
+import { createStagingAreaForInvariantContext } from './misc'
+
+import type {
+  PickUpTipRunTimeCommand,
+  ProtocolAnalysisOutput,
+  RunTimeCommand,
+} from '@opentrons/shared-data'
+import type {
+  InvariantContext,
+  LabwareEntities,
+  ModuleEntities,
+  PipetteEntities,
+  StagingAreaEntities,
+  TrashBinEntities,
+  WasteChuteEntities,
+} from '../types'
+
+export function constructInvariantContextFromAnalysis(
+  analysis: ProtocolAnalysisOutput
+): InvariantContext {
+  const { labware, modules, pipettes, commands } = analysis
+  const labwareDefinitions = getLabwareDefinitionsByURIForProtocol(commands)
+
+  const moduleEntities: ModuleEntities = modules.reduce(
+    (acc: ModuleEntities, module) => {
+      const { id, model } = module
+      acc[id] = {
+        id,
+        type: getModuleType(model),
+        model,
+        pythonName: 'n/a',
+      }
+
+      return acc
+    },
+    {}
+  )
+
+  const labwareEntities: LabwareEntities = labware.reduce(
+    (acc: LabwareEntities, loadedLabware) => {
```

Comment:
Should we type like this?
```suggestion
  const labwareEntities: LabwareEntities = labware.reduce<LabwareEntities>(
    (acc, loadedLabware) => {
```